### ISSUE TITLE
Proposal: group syscalls with arguments + support grouping in both LHS and RHS

### ISSUE THREAD

**Issue by andronat** [daniel-grumberg/varan-div#18]
It would be very convenient to allow grouping syscalls with common arguments. Supporting FD ignoring requires multiple syscalls that accept an FD argument, to be handled homogeneously.

Example:
```
nothing => fstat(fd,_) {
  return is_ignored_file_fd($(fd));
}

nothing => fdatasync(fd) {
  return is_ignored_file_fd($(fd));
}

nothing => read(fd,_,_) {
  return is_ignored_file_fd($(fd));
}

nothing => write(fd,_,_) {
  return is_ignored_file_fd($(fd));
}
```

Instead we could support this:
```
group fd_calls => [read(fd,_,_), write(fd,_,_), fdatasync(fd), fstat(fd,_)]

nothing => fd_calls { return is_ignored_file_fd($(fd)); }
```

**Comment by luisggpina** [comment]
Apart from ensuring that all the rules in the group have an argument called `fd`, implementing this should be trivial.

**Comment by daniel-grumberg** [comment]
Groups should not be allowed on the right hand side apart from the nothing case ... because otherwise it breaks the idea that rules for a given left-hand side should match the same thing. Imagine the following example:
```
group fd_calls => [read(fd,_,_), write(fd,_,_), fdatasync(fd), fstat(fd,_)]

getpid() => fd_calls,
                 self()
```
The problem is that now we have multiple rules for `getpid` that are not predicated. This is fine if we completely change the semantics of group and instead of making a preprocessor kind of thing, we implement it at the code generation level.
That is instead of making group mean: "just create a bunch of rules where you replace the group identifier with the syscall values", the semantics would be: "anyone of those system calls is acceptable at this point in the sequence"
Thoughts?

**Comment by luisggpina** [comment]
I think the multiple rules for `getpid` in this case are acceptable.  I don't understand the semantics change you point out, though, perhaps I need a better example.

In the case you showed, I would expect this to be exactly equivalent to the following file:
```
getpid() => read(fd,_,_),
                 self()

getpid() => write(fd,_,_),
                 self()

getpid() => fdatasync(fd),
                 self()

getpid() => fstat(fd,_),
                 self()
```

And then follow the existing rules to match system call rules.

Is this OK, or am I missing something subtle?

**Comment by daniel-grumberg** [comment]
Yes, but there is no way to differentiate any of those rules using just the left-hand side, thus this needs to be lifted into the language and handled at "runtime".
